packages/snaps-sdk/src/types/caip.ts

```diff
@@ -22,3 +34,110 @@ export type ChainId = CaipChainId;
  * @deprecated Use {@link CaipAccountId} instead.
  */
 export type AccountId = CaipAccountId;
+
+/**
+ * A stuct representing a list of CAIP-10 account IDs where the account addresses and namespaces are the same.
+ */
+export const CaipAccountIdsMatchedByAddressAndNamespaceStruct = refine(
+  array(CaipAccountIdStruct),
+  'Matching Addresses Account ID List',
+  (value) => {
+    const parsedAccountIds = value.map((accountId) =>
+      parseCaipAccountId(accountId),
+    );
+
+    if (
+      !parsedAccountIds.every(
+        ({ address, chain: { namespace } }) =>
+          address === parsedAccountIds[0].address &&
+          namespace === parsedAccountIds[0].chain.namespace,
+      )
+    ) {
+      return 'All account IDs must have the same address and namespace.';
+    }
+
+    return true;
+  },
+);
+
+/**
+ * A list of CAIP-10 account IDs where the account addresses are the same.
+ */
+export type MatchingAddressesCaipAccountIdList = Infer<
+  typeof CaipAccountIdsMatchedByAddressAndNamespaceStruct
+>;
+
+/**
+ * A struct representing a list of non-EIP-155 CAIP-10 account IDs where the account addresses are the same.
+ */
+export const NonEip155CaipAccountIdsMatchedByAddressAndNamespaceStruct = refine(
+  CaipAccountIdsMatchedByAddressAndNamespaceStruct,
+  'Non-EIP-155 Matching Addresses Account ID List',
+  (value) => {
+    const containsEip155 = value.some((accountId) => {
+      const {
+        chain: { namespace },
+      } = parseCaipAccountId(accountId);
+
+      return namespace === KnownCaipNamespace.Eip155;
+    });
+
+    if (containsEip155) {
+      return 'All account IDs must have non-EIP-155 namespaces.';
+    }
+    return true;
+  },
+);
+
+/**
+ * A list of non-EIP-155 CAIP-10 account IDs where the account addresses are the same.
+ */
+export type NonEip155MatchingAddressesCaipAccountIdList = Infer<
```

Comment:
Hmm, the types here will probably be the same as existing types we have for CAIP-10. I'm wondering if its cleaner to re-use the types and only introduce new structs to validate the additional restrictions?

WDYT @Mrtenz ?

Reply:
Do you mean using `CaipAccountId` in the type and `NonEip155MatchingAddressesCaipAccountIdList` in validation ?

I thought about it as this type isn't directly explicit but did not update it for consistency but I agree it would be cleaner to just use `CaipAccountId`

Reply:
Yeah, because the type will likely be exactly the same, it doesn't include the struct refinements